sklearn/linear_model/ridge.py

```diff
@@ -1001,18 +1079,164 @@ def _diag_dot(self, D, B):
             D = D[(slice(None), ) + (np.newaxis, ) * (len(B.shape) - 1)]
         return D * B
 
-    def _errors_and_values_helper(self, alpha, y, v, Q, QT_y):
-        """Helper function to avoid code duplication between self._errors and
-        self._values.
+    def _compute_gram(self, X, sqrt_sw):
+        """Computes the Gram matrix with possible centering.
 
-        Notes
-        -----
-        We don't construct matrix G, instead compute action on y & diagonal.
+        If ``center`` is ``True``, compute
+        (X - X.mean(axis=0)).dot((X - X.mean(axis=0)).T)
+        else X.dot(X.T)
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            The input uncentered data.
+
+        sqrt_sw : ndarray, shape (n_samples,)
+            square roots of sample weights
+
+        center : bool, default is True
+            Whether or not to remove the mean from ``X``.
+
+        Returns
+        -------
+        gram : ndarray, shape (n_samples, n_samples)
+            The Gram matrix.
+        X_mean : ndarray, shape (n_feature,)
+            The mean of ``X`` for each feature.
+        """
+        center = self.fit_intercept and sparse.issparse(X)
+        if not center:
+            # in this case centering has been done in preprocessing
+            # or we are not fitting an intercept.
+            X_mean = np.zeros(X.shape[1], dtype=X.dtype)
+            return safe_sparse_dot(X, X.T, dense_output=True), X_mean
+        # otherwise X is always sparse
+        n_samples = X.shape[0]
+        sample_weight_matrix = sparse.dia_matrix(
+            (sqrt_sw, 0), shape=(n_samples, n_samples))
+        X_weighted = sample_weight_matrix.dot(X)
+        X_mean, _ = mean_variance_axis(X_weighted, axis=0)
+        X_mean *= n_samples / sqrt_sw.dot(sqrt_sw)
+        X_mX = sqrt_sw[:, None] * safe_sparse_dot(
+            X_mean, X.T, dense_output=True)
+        X_mX_m = np.empty((n_samples, n_samples), dtype=X.dtype)
+        X_mX_m[:, :] = np.dot(X_mean, X_mean)
```

Comment:
```python
        X_mX_m = np.full((n_samples, n_samples),
                         fill_value=np.dot(X_mean, X_mean), dtype=X.dtype)
```

Reply:
Actually, why not just:

```python
X_mX_m = np.dot(X_mean, X_mean)
```

or if dtype is not guaranteed to match:

```python
X_mX_m = np.dot(X_mean, X_mean).astype(X.dtype, copy=False)
```